### DISCUSSION TITLE
Custom Module Links+Navigation

### DISCUSSION THREAD

**Discussion by ernst86913** [directus/directus#5468] [Feature Requests]
Possibility to add MAIN MENU entries for standard collections without adding custom modules.

Status quo:   Currently all Data Models are presented in "Collections"
Improvement: Assign Collections to different MAIN MENU entries.

Is there currently any work in progress?  Is this something you are considering?

**Example Menu (HR, Finance, Projects, Marketing, Sales should be main menu items)**

* HR
---- Employees
---- Contractors
---- Recruiting
---- ....
* Finance
---- Invoices
---- Expenses
---- Contracts
---- Accounting
---- ....
* Projects
---- Portfolio
---- Time Booking
---- Reports
---- ....
* Marketing
---- Target Segments
---- Campaigns
---- ....
* Sales
---- Customers
---- Prospects
----  ...

**Comment by rijkvanzanten** [comment]
This feels very similar to the Collection Navigation option under Roles & Permissions 🙂

**Reply by ernst86913** [reply]
... maybe I got it wrong... but as I understand it this option just allows to have menu items under the main module "Collection"... but it does not allow to have a new main module entry in the main module...

Or is there a way to do it there?

**Reply by ernst86913** [reply]
(I have a long list of menu items... putting them all under ONE main item (=Collections) does not make sense from usability perspective...

**Reply by rijkvanzanten** [reply]
It allows you to make multiple groups of collections, each with a title and optionally collapsable. Sounds like what you're describing above 🙂

**Reply by ernst86913** [reply]
Close - but not 100%.... I'd like to use this functionality... but also have the option to have collections in a separate MAIN menu entry... and not all of them in "Collections". :-)    But... without creating a completely separate module for it...

![image](https://user-images.githubusercontent.com/47870436/117085124-3b0f2c00-ad49-11eb-8f12-f0f4f839e2da.png)

**Reply by rijkvanzanten** [reply]
Ohhhhhhhh the Module Navigation, gotcha 👍🏻

**Reply by ernst86913** [reply]
Yes - sorry for confusion :-)

**Comment by benhaynes** [comment]
This is not possible without custom extensions. However, I could imagine a reusable custom Module that allows you to configure placeholder icon/name/collections/navigations. This single extension would allow for endlessly customizing top-level modules that serve as bookmarks to other pages.

**Reply by ernst86913** [reply]
Well... guess the perfect solution would be to introduce a "Menu Editor" as part of the settings... there you can define the menu hierarchy (Modules / Links/Dividers/Groups,....). And in the role permissions you can define for each whether the role can see this entry or not... this would also allow future multi-role assignments to users. If you have completely independent menus per user role you can't merge the menu as part of multi-role assignments.

Tree in project settings (as JSON?):
MODULE (icon, name, optional link if the sub-menu is provided directly by the module)
optional MODULE-MENU ENTRIES if sub-menu is not provided directly by module   (icon, name, divider, groups,...)

And in the permissions the full tree is shown for each role... and allows for selection what is visible for the role and what not... permissions could be saved in the project settings JSON above.

Alternative would be a separate directus_menu table

Question is whether this is something you want to touch right now.... and whether you consider this as a needed feature as part of your 80/20 rule...

**Reply by ernst86913** [reply]
BTW - I will develop this for my purposes anyway... so if you give me guidance on whether/how you would like to have it I can follow that accordingly.....

**Reply by benhaynes** [reply]
Thanks! This would most likely not be something that is 80/20, which is why I mentioned the custom extensions. However, all of this could be handled by a pretty simple "dynamic" extension... so it can be re-used multiple times. As of now, in Directus Core you can adjust the Module Nav (just enable/disable) and Collection Nav (inside the Collection Module) that includes grouping headers. Adding _new_ modules is done through Module extensions... and I don't see a strong reason to mix those between Modules and another level of Collection link organization.

**Reply by ernst86913** [reply]
@benhaynes ... just had a look at it... there might be even an easier approach to this... use the existing "ADD MODULE" functionality... and within it have 2 options:   a)  CUSTOM MODULE with a link as its done right now   b) COLLECTION MODULE where you can select collections that should be part of it... this would be exactly the same approach as with the existing COLLECTION Navigation... should be straight forward.... what do you think about this approach?

**Reply by benhaynes** [reply]
I think that would certainly work as a a configuration within our current setup — my only concern is if this is actually 80/20, and useful enough for the community to justify the added complexity/code. Let's see if this Discussion gets some attention and 👍 s.

**Reply by ernst86913** [reply]
@benhaynes  As I am just digging into how the custom module for modules/navigation configuration could work I have one essential question: where should such settings be stored without affecting the current data model? Best approach that comes to my mind would be storing it as a setting in the "SettingsStore". But - the SettingsStore is currently written in a way that it does not allow extending it.

a) Whats from your perspective the right place to store module settings?
b) Wouldn't it make sense to change the way the SettingsStore works to allow multiple rows per project?
e.g. PROJECT NAME, SETTING NAME, JSON ... and use SETTING NAME to identify the right setting within the project... and use the JSON for storing any settings of any format?
c) Whats the right place from a UI perspective to make module settings available to customers?

But maybe I am missing something....

**Reply by benhaynes** [reply]
I'm not sure about Settings for Modules... that would require a more technical response from @rijkvanzanten.

**Reply by ernst86913** [reply]
@rijkvanzanten can you please help me with that one? Thx

**Reply by rijkvanzanten** [reply]
There's no native way for modules to do database modifications or anything like that. Migrations are already an available custom migration type, so I think we'll rely on that for modules to be able to setup their own tables if need be.

> Wouldn't it make sense to change the way the SettingsStore works to allow multiple rows per project?

No it doesn't. That's how it worked before (< v9) and came with a range of issues and hacks that made the tradeoff for this kinda usage not worth it. For starters, you lose data types completely, which is impossible to properly mitigate on the JS layer, you lose relationships, as every column is a flat string, migrations became non standard, as we now had to look for a row instead of a known column by name, etc

**Reply by ernst86913** [reply]
Thanks a lot... though not sure how migrations can be used to do that... as I understand migrations those are dependent on the directus version... and not on which modules are installed.

**Reply by ernst86913** [reply]
@rijkvanzanten Just figured out that I can anyways create a custom field in project settings where I can store the menu structure... so - should be fine... question then is whats the guidance for modules to setup tables / fields... and clean them up when a module is de-installed again... I assume should be similar to directus standard approach of having YAML files for that purpose.

**Reply by tarwin** [reply]
@ernst86913 would love to test / work on the extension with you. Looking for similar.

**Comment by danieljwestman** [comment]
I have previously built a PoC for kind of something I think you are looking for @ernst86913 

My idea was to build **two custom modules**: `collection` and `collection_categories` (the naming could maybe be better, yes).

**The first one** links to a specific collection of choice (the path would be `/collection/[collection name]`). In the collection navigation bar you could have the items in that particular collection (best suited for collections that haven't that much items, or you could build some kind of pagination).

**The second one** links to a set of collections based on a custom field `category`. This one works just like the core module `collections` but filters out based on a particular category (the path would be `/collection_categories/[category name]`).

These two custom modules can be used for multiple navigation options in the module bar, obviously, and result in a nice custom navigation experience.

It all worked pretty good but I didn't complete the implementation of this PoC. I'm planning to go for it again in a couple of weeks. It depends a bit on how the extensions infrastructure will be formed in core (there are a few loose ends and different ideas here).

Maybe this can inspire you ☺️

**Comment by oreilles** [comment]
Hey, loving this idea, I would also love to be able to set different level of hierarchy in the collection grouping. 👍 
However I think you're going the wrong way about it. As @rijkvanzanten said, the grouping option already exists, I'm sure there would be a way to implement different level of nesting within it, so that it could be displayed kind of like the select dropdown (which also supports arbitrary nesting levels). Will look into it.

**Comment by tarwin** [comment]
I had this in a data-centric CMS I made many years ago. May give some ideas.

From 9min -> 11min it shows how this idea works:
https://vimeo.com/7428887

**Reply by tarwin** [reply]
One thing to note it would be good to be able to set bookmarks as items in a list like this.

For example, you might have a `Posts` collection, which has two types: `Blog` and `Press`. You may then want to have filtered bookmarks added to the menu so you would end up with a menu:

```
- Blog
- Press
```

As an addition (not only related to this) you could add enforced fields values to filtered items like this so when you add a new `Blog` it's type is set to `blog`, or at least defaults to this.

**Comment by br41nslug** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
